nntrainer/dataset/raw_file_data_producer.cpp

```diff
@@ -0,0 +1,180 @@
+
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file   raw_file_data_producer.cpp
+ * @date   12 July 2021
+ * @brief  This file contains raw file data producers, reading from a file
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+
+#include <raw_file_data_producer.h>
+
+#include <fstream>
+#include <memory>
+#include <numeric>
+#include <random>
+#include <vector>
+
+#include <base_properties.h>
+#include <nntrainer_error.h>
+#include <node_exporter.h>
+#include <util_func.h>
+
+namespace nntrainer {
+
+/**
+ * @brief Props containing file path value
+ *
+ */
+class PropsPath : public Property<std::string> {
+public:
+  static constexpr const char *key = "path"; /**< unique key to access */
+  using prop_tag = str_prop_tag;             /**< property type */
+
+  bool isValid(const std::string &v) {
+    std::ifstream file(v, std::ios::binary | std::ios::ate);
+    return file.good();
+  }
+
+  void set(const std::string &v) {
+    Property<std::string>::set(v);
+    std::ifstream file(v, std::ios::binary | std::ios::ate);
+    cached_pos_size = file.tellg();
+  }
+
+  std::ifstream::pos_type file_size() { return cached_pos_size; }
+
+private:
+  std::ifstream::pos_type cached_pos_size;
+};
+
+RawFileDataProducer::RawFileDataProducer() : raw_file_props(new PropTypes()) {}
+RawFileDataProducer::~RawFileDataProducer() {}
+
+const std::string RawFileDataProducer::getType() const {
+  return RawFileDataProducer::type;
+}
+
+unsigned long long
+RawFileDataProducer::size(const std::vector<TensorDim> &input_dims,
+                          const std::vector<TensorDim> &label_dims) const {
+  auto size_accumulator = [](const unsigned int &a, const TensorDim &b) {
+    return a + b.getFeatureLen();
+  };
+
+  auto sample_size =
+    std::accumulate(input_dims.begin(), input_dims.end(), 0u, size_accumulator);
+  sample_size = std::accumulate(label_dims.begin(), label_dims.end(),
+                                sample_size, size_accumulator);
+
+  auto path_prop = std::get<PropsPath>(*raw_file_props);
+  auto file_size = path_prop.file_size();
+
+  /// checking alignment is a good way to make check if a file is valid,
+  /// unfortunately, our test dataset does not have this property
+  /// (trainingSet.dat, valSet.dat, testSet.dat) after checking, we can
+  /// uncomment below line.
+  // NNTR_THROW_IF((file_size % sample_size * RawFileDataProducer::pixel_size !=
+  // 0),
+  //               std::invalid_argument)
+  //   << " Given file does not align with the given sample size, sample size: "
+  //   << sample_size << " file_size: " << file_size;
+
+  return path_prop.file_size() /
```

Comment:
return file_size / ...

Reply:
why, wonder how it worked.

Reply:
There is a file size right below. clang-format doesn't allow me to hoist the below line :p

Reply:
ok.